src/base/filesystemwatcher.cpp

```diff
@@ -1,145 +1,128 @@
+
+#include "base/algorithm.h"
 #include "base/bittorrent/magneturi.h"
 #include "base/bittorrent/torrentinfo.h"
+#include "base/global.h"
 #include "base/preferences.h"
-#include "filesystemwatcher.h"
-
-#ifndef CIFS_MAGIC_NUMBER
-#define CIFS_MAGIC_NUMBER 0xFF534D42
-#endif
-
-#ifndef NFS_SUPER_MAGIC
-#define NFS_SUPER_MAGIC 0x6969
-#endif
-
-#ifndef SMB_SUPER_MAGIC
-#define SMB_SUPER_MAGIC 0x517B
-#endif
+#include "base/utils/fs.h"
 
-const int WATCH_INTERVAL = 10000; // 10 sec
-const int MAX_PARTIAL_RETRIES = 5;
+namespace
+{
+    const int WATCH_INTERVAL = 10000; // 10 sec
+    const int MAX_PARTIAL_RETRIES = 5;
+}
 
 FileSystemWatcher::FileSystemWatcher(QObject *parent)
     : QFileSystemWatcher(parent)
 {
-    m_filters << "*.torrent" << "*.magnet";
-    connect(this, SIGNAL(directoryChanged(QString)), SLOT(scanLocalFolder(QString)));
-}
+    connect(this, &QFileSystemWatcher::directoryChanged, this, &FileSystemWatcher::scanLocalFolder);
+
+    m_partialTorrentTimer.setSingleShot(true);
+    connect(&m_partialTorrentTimer, &QTimer::timeout, this, &FileSystemWatcher::processPartialTorrents);
 
-FileSystemWatcher::~FileSystemWatcher()
-{
 #ifndef Q_OS_WIN
-    if (m_watchTimer)
-        delete m_watchTimer;
+    connect(&m_watchTimer, &QTimer::timeout, this, &FileSystemWatcher::scanNetworkFolders);
 #endif
-    if (m_partialTorrentTimer)
-        delete m_partialTorrentTimer;
 }
 
 QStringList FileSystemWatcher::directories() const
 {
-    QStringList dirs;
+    QStringList dirs = QFileSystemWatcher::directories();
 #ifndef Q_OS_WIN
-    if (m_watchTimer) {
-        foreach (const QDir &dir, m_watchedFolders)
-            dirs << dir.canonicalPath();
-    }
+    for (const QDir &dir : qAsConst(m_watchedFolders))
+        dirs << dir.canonicalPath();
 #endif
-    dirs << QFileSystemWatcher::directories();
     return dirs;
 }
 
 void FileSystemWatcher::addPath(const QString &path)
 {
+    if (path.isEmpty()) return;
+
 #if !defined Q_OS_WIN && !defined Q_OS_HAIKU
-    QDir dir(path);
+    const QDir dir(path);
     if (!dir.exists()) return;
 
     // Check if the path points to a network file system or not
-    if (isNetworkFileSystem(path)) {
+    if (Utils::Fs::isNetworkFileSystem(path)) {
         // Network mode
         qDebug("Network folder detected: %s", qUtf8Printable(path));
         qDebug("Using file polling mode instead of inotify...");
         m_watchedFolders << dir;
-        // Set up the watch timer
-        if (!m_watchTimer) {
-            m_watchTimer = new QTimer(this);
-            connect(m_watchTimer, SIGNAL(timeout()), SLOT(scanNetworkFolders()));
-            m_watchTimer->start(WATCH_INTERVAL); // 5 sec
-        }
-    }
-    else {
-#endif
-        // Normal mode
-        qDebug("FS Watching is watching %s in normal mode", qUtf8Printable(path));
-        QFileSystemWatcher::addPath(path);
-        scanLocalFolder(path);
-#if !defined Q_OS_WIN && !defined Q_OS_HAIKU
+
+        m_watchTimer.start(WATCH_INTERVAL);
```

Comment:
The same question: what about already active timer?

Reply:
I think checking status of `m_watchTimer` can be optional, assuming users won't be continuously adding path all the time.